From dec5a356627ed59975336fc502623b938d39a944 Mon Sep 17 00:00:00 2001
From: Raphael Kim <raphk@google.com>
Date: Thu, 2 Jun 2022 18:14:31 +0000
Subject: [PATCH] Extract app label from component name in notification access
 confirmation UI

Bug: 228178437
Test: Manually tested on POC
Change-Id: I2df1a0e82c74a02e3214ca8fc9b44e14a62ee622
Merged-In: I4c4fbca57e3e52b5d20aee6144d3a1be3c5352c2
(cherry picked from commit 728fda8ca7eabc7b5ef93b7f3c572c9c1e80b6e7)
Merged-In: I2df1a0e82c74a02e3214ca8fc9b44e14a62ee622
---
 ...otificationAccessConfirmationActivity.java | 33 ++++++++++++++++---
 1 file changed, 29 insertions(+), 4 deletions(-)

diff --git a/tests/CarDeveloperOptions/src/com/android/car/developeroptions/notification/NotificationAccessConfirmationActivity.java b/tests/CarDeveloperOptions/src/com/android/car/developeroptions/notification/NotificationAccessConfirmationActivity.java
index b7696bf25..9f6409728 100644
--- a/tests/CarDeveloperOptions/src/com/android/car/developeroptions/notification/NotificationAccessConfirmationActivity.java
+++ b/tests/CarDeveloperOptions/src/com/android/car/developeroptions/notification/NotificationAccessConfirmationActivity.java
@@ -18,7 +18,6 @@
 package com.android.car.developeroptions.notification;
 
 import static com.android.internal.notification.NotificationAccessConfirmationActivityContract.EXTRA_COMPONENT_NAME;
-import static com.android.internal.notification.NotificationAccessConfirmationActivityContract.EXTRA_PACKAGE_TITLE;
 import static com.android.internal.notification.NotificationAccessConfirmationActivityContract.EXTRA_USER_ID;
 
 import android.Manifest;
@@ -28,10 +27,13 @@ import android.app.NotificationManager;
 import android.content.ComponentName;
 import android.content.Context;
 import android.content.DialogInterface;
+import android.content.pm.ApplicationInfo;
+import android.content.pm.PackageItemInfo;
 import android.content.pm.PackageManager;
 import android.content.pm.ServiceInfo;
 import android.os.Bundle;
 import android.os.UserHandle;
+import android.text.TextUtils;
 import android.util.Slog;
 import android.view.WindowManager;
 import android.view.accessibility.AccessibilityEvent;
@@ -59,15 +61,38 @@ public class NotificationAccessConfirmationActivity extends Activity
 
         mComponentName = getIntent().getParcelableExtra(EXTRA_COMPONENT_NAME);
         mUserId = getIntent().getIntExtra(EXTRA_USER_ID, UserHandle.USER_NULL);
-        String pkgTitle = getIntent().getStringExtra(EXTRA_PACKAGE_TITLE);
+        CharSequence mAppLabel;
+
+        if (mComponentName == null || mComponentName.getPackageName() == null) {
+            finish();
+            return;
+        }
+
+        try {
+            ApplicationInfo applicationInfo = getPackageManager().getApplicationInfo(
+                    mComponentName.getPackageName(), 0);
+            mAppLabel = applicationInfo.loadSafeLabel(getPackageManager(),
+                    PackageItemInfo.DEFAULT_MAX_LABEL_SIZE_PX,
+                    PackageItemInfo.SAFE_LABEL_FLAG_TRIM
+                            | PackageItemInfo.SAFE_LABEL_FLAG_FIRST_LINE);
+        } catch (PackageManager.NameNotFoundException e) {
+            Slog.e(LOG_TAG, "Couldn't find app with package name for " + mComponentName, e);
+            finish();
+            return;
+        }
+
+        if (TextUtils.isEmpty(mAppLabel)) {
+            finish();
+            return;
+        }
 
         AlertController.AlertParams p = new AlertController.AlertParams(this);
         p.mTitle = getString(
                 R.string.notification_listener_security_warning_title,
-                pkgTitle);
+                mAppLabel);
         p.mMessage = getString(
                 R.string.notification_listener_security_warning_summary,
-                pkgTitle);
+                mAppLabel);
         p.mPositiveButtonText = getString(R.string.allow);
         p.mPositiveButtonListener = (a, b) -> onAllow();
         p.mNegativeButtonText = getString(R.string.deny);
-- 
2.37.0.rc0.161.g10f37bed90-goog

